system-configuration/src/network_configuration.rs

```diff
@@ -0,0 +1,468 @@
+// Copyright 2017 Amagicom AB.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Bindings to [`SCNetworkConfiguration`].
+//!
+//! See the examples directory for examples how to use this module.
+//!
+//! [`SCNetworkConfiguration`]: https://developer.apple.com/documentation/systemconfiguration/scnetworkconfiguration
+
+use core_foundation::array::CFArray;
+use core_foundation::base::{CFType, TCFType};
+use core_foundation::base::kCFAllocatorDefault;
+use core_foundation::dictionary::CFDictionary;
+use core_foundation::string::{CFString, CFStringRef};
+
+use dynamic_store::{SCDynamicStore, SCDynamicStoreBuilder};
+pub use system_configuration_sys::network_configuration::*;
+use system_configuration_sys::preferences::SCPreferencesCreate;
+
+use std::{fmt, ptr};
+use std::net::IpAddr;
+
+/// MTU
+#[derive(Debug)]
+pub struct SCNetworkInterfaceMtu {
+    /// the current MTU setting for the interface.
+    pub current: u32,
+    /// the minimum MTU setting for the interface. If None, the minimum setting could not
+    /// be determined.
+    pub min: Option<u32>,
+    /// the maximum MTU setting for the interface. If None, the maximum setting could not
+    /// be determined.
+    pub max: Option<u32>,
+}
+
+/// DNS
+#[derive(Debug)]
+pub struct SCNetworkServiceDns {
+    state_domain_name: Option<String>,
+    setup_domain_name: Option<String>,
+    state_server_addresses: Option<Vec<IpAddr>>,
+    setup_server_addresses: Option<Vec<IpAddr>>,
```

Comment:
>  Note that I made all fields public again. Since these are pure data structs I think we can just allow direct access. I also made Option<Vec<IpAddr>> into Vec<IpAddr>, I thought we could just model no addresess as an empty vector instead of a None, would that have downsides?

Personally, I think `Option<Vec<IpAddr>>` is better than `Vec<IpAddr>`, when we setting dns with `Empty` (like `networksetup -setdnsservers "Wi-Fi" "Empty"`), we do not need alloc one `Vec` .
But if you insist on that, i can change that.

Reply:
Creating an empty Vec does not do any allocation in Rust unless I'm remembering wrong. So I would definitely not worry about the performance aspect of it.

But yes if we want to be able to detect the difference of no DNS set vs empty list of DNS set, then we need this to be an `Option<..>`. So keep it an `Option<..>`, that is then also same as the type of argument to `set_dns_state_addresses` method, so that is good.